File: golang/day_01/day_01.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"
)

func readFile(filename string) []string {
	actualFilePath := filepath.Join("..", "..", "inputs", "day_01", filename)
	dat, err := os.ReadFile(actualFilePath)

	if err != nil {
		panic(err)
	}

	stringDat := string(dat)

	stringDat = strings.ReplaceAll(stringDat, "\r\n", "\n")

	return strings.Split(stringDat, "\n")
}

func processLine(line string) int {
	chars := strings.Split(line, "")

	var firstDigit, lastDigit int
	firstDigit = -1

	for _, char := range chars {
		digit, err := strconv.Atoi(char)

		if err != nil {
			continue
		}

		if firstDigit == -1 {
			firstDigit = digit
		}

		lastDigit = digit
	}

	return firstDigit*10 + lastDigit
}

func part_one() {
	readData := readFile("input.txt")

	var total int

	for _, line := range readData {
		if line == "" {
			continue
		}
		current := processLine(line)
		total += current
	}

	fmt.Println(total)
}

func processLinePartTwo(line string) int {
	line = strings.ReplaceAll(line, "one", "o1e")
	line = strings.ReplaceAll(line, "two", "t2o")
	line = strings.ReplaceAll(line, "three", "t3e")
	line = strings.ReplaceAll(line, "four", "f4r")
	line = strings.ReplaceAll(line, "five", "f5e")
	line = strings.ReplaceAll(line, "six", "s6x")
	line = strings.ReplaceAll(line, "seven", "s7n")
	line = strings.ReplaceAll(line, "eight", "e8t")
	line = strings.ReplaceAll(line, "nine", "n9e")

	return processLine(line)
}

func part_two() {
	readData := readFile("input.txt")

	var total int

	for _, line := range readData {
		if line == "" {
			continue
		}
		current := processLinePartTwo(line)
		total += current
	}

	fmt.Println(total)
}

func main() {
	part_one()
	part_two()
}

File: elixir/day_01/day_01.ex
defmodule Day01 do
  defp read_input(input) do
    Path.join(["..", "..", "inputs", "day_01", input])
    |> File.read!()
    |> String.replace("\r\n", "\n")
    |> String.split("\n")
    |> Enum.filter(&(&1 != ""))
  end

  defp first_and_last(list) do
    [Enum.at(list, 0), Enum.at(list, -1)]
  end

  defp process_line(line) do
    line
    # Get each character
    |> String.graphemes()
    # Filter out non-numbers
    |> Enum.filter(fn x -> String.match?(x, ~r/^-?\d+$/) end)
    # Convert to integers
    |> Enum.map(&String.to_integer/1)
    |> first_and_last()
    # Convert to a single number
    |> Enum.reduce(&(&1 + &2 * 10))
  end

  def part_one do
    read_input("input.txt")
    |> Enum.map(&process_line/1)
    |> Enum.sum()
  end

  defp process_line_two(line) do
    line
    |> String.replace("one", "o1e")
    |> String.replace("two", "t2o")
    |> String.replace("three", "t3e")
    |> String.replace("four", "f4r")
    |> String.replace("five", "f5e")
    |> String.replace("six", "s6x")
    |> String.replace("seven", "s7n")
    |> String.replace("eight", "e8t")
    |> String.replace("nine", "n9e")
    |> process_line()
    |> IO.inspect()
  end

  def part_two do
    read_input("input.txt")
    |> Enum.map(&process_line_two/1)
    |> Enum.sum()
  end
end

Day01.part_two() |> IO.inspect()
